client/src/pages/Login.tsx

```diff
@@ -0,0 +1,130 @@
+import React, { useEffect, useState } from "react";
+import { fetchJson, fetchText } from "./../utils";
+import { TextInputField } from "./../components/Components";
+import { Container, Grid, Typography, Button } from "@material-ui/core";
+import { IdentityToken, emptyIdentityToken } from "./../interfaces";
+import Logo from "./../img/Logo.png";
+import "./../css/Login.css";
+
+// Applease typescript
+declare const window: any;
+// Grab handle to our microapps js library
+const microapps = window.microapps;
+
+function Login() {
+  const is_web = window.location == window.parent.location;
+  const is_debug = true;
+  const speed = 3;
+
+  // Flag for us to manually run login on mobile
+  const [loginError, setLoginError] = useState(false);
+  // User Identity
+  const [identity, setIdentity] = useState<IdentityToken>(emptyIdentityToken());
+
+  const updateUser = (text: string) => {
+    const newIdentity = emptyIdentityToken();
+    newIdentity.sub = text;
+    setIdentity(newIdentity);
+  };
+
+  const login = () => {
+    console.log("Login type: " + (is_web ? "Web" : "Microapp"));
+    if (is_web) {
+      // Web 'faked' stuff
+    } else {
+      const request = { nonce: "Don't Hack me please" };
+      microapps
+        .getIdentity(request)
+        .then((response: any) => {
+          const decoded = JSON.parse(atob(response.split(".")[1]));
```

Comment:
Same as above comment.
As you have used this for multiple times, consider extracting it into some common library

Reply:
abstracted into function in utils